store/tikv/coprocessor.go

```diff
@@ -456,84 +465,158 @@ func (it *copIterator) Next() ([]byte, error) {
-			err = bo.Backoff(BoRegionMiss, errors.New(regionErr.String()))
-			if err != nil {
-				return []copResponse{{err: errors.Trace(err)}}
-			}
-			return it.handleRegionErrorTask(bo, task)
+		// We may meet RegionError at the first packet, but not during visiting the stream.
+		coprocessorCounter.WithLabelValues("rebuild_task").Inc()
+		return buildCopTasks(bo, it.store.regionCache, task.ranges, it.req.Desc, it.req.Streaming)
+	}
+	if lockErr := resp.GetLocked(); lockErr != nil {
+		log.Debugf("coprocessor encounters lock: %v", lockErr)
+		ok, err1 := it.store.lockResolver.ResolveLocks(bo, []*Lock{NewLock(lockErr)})
+		if err1 != nil {
+			return nil, errors.Trace(err1)
 		}
-		if e := resp.Cop.GetLocked(); e != nil {
-			log.Debugf("coprocessor encounters lock: %v", e)
-			ok, err1 := it.store.lockResolver.ResolveLocks(bo, []*Lock{NewLock(e)})
-			if err1 != nil {
-				return []copResponse{{err: errors.Trace(err1)}}
-			}
-			if !ok {
-				err = bo.Backoff(boTxnLockFast, errors.New(e.String()))
-				if err != nil {
-					return []copResponse{{err: errors.Trace(err)}}
-				}
+		if !ok {
+			if err := bo.Backoff(boTxnLockFast, errors.New(lockErr.String())); err != nil {
+				return nil, errors.Trace(err)
 			}
-			continue
-		}
-		if e := resp.Cop.GetOtherError(); e != "" {
-			err = errors.Errorf("other error: %s", e)
-			log.Warnf("coprocessor err: %v", err)
-			return []copResponse{{err: errors.Trace(err)}}
 		}
-		task.storeAddr = sender.storeAddr
-		return []copResponse{{Response: resp.Cop}}
+		return buildCopTasksFromRemain(bo, it.store.regionCache, resp, task, it.req.Desc, it.req.Streaming)
 	}
+	if otherErr := resp.GetOtherError(); otherErr != "" {
+		err := errors.Errorf("other error: %s", otherErr)
+		log.Warnf("coprocessor err: %v", err)
+		return nil, errors.Trace(err)
+	}
+	return nil, nil
 }
 
-// handleRegionErrorTask handles current task. It may be split into multiple tasks (in region split scenario).
-func (it *copIterator) handleRegionErrorTask(bo *Backoffer, task *copTask) []copResponse {
-	coprocessorCounter.WithLabelValues("rebuild_task").Inc()
-
-	newTasks, err := buildCopTasks(bo, it.store.regionCache, task.ranges, it.req.Desc)
-	if err != nil {
-		return []copResponse{{err: errors.Trace(err)}}
-	}
-	if newTasks == nil {
-		// TODO: check this, this should never happen.
-		return nil
+func buildCopTasksFromRemain(bo *Backoffer, cache *RegionCache, resp *coprocessor.Response, task *copTask, desc bool, streaming bool) ([]*copTask, error) {
+	if !streaming {
+		return buildCopTasks(bo, cache, task.ranges, desc, streaming)
 	}
 
-	ret := make([]copResponse, 0, len(newTasks))
-	for _, t := range newTasks {
-		resp := it.handleTask(bo, t)
-		ret = append(ret, resp...)
+	ranges := calculateRemain(task.ranges, resp.Range, desc)
+	return buildCopTasks(bo, cache, ranges, desc, streaming)
+}
+
+func calculateRemain(ranges *copRanges, split *coprocessor.KeyRange, desc bool) *copRanges {
```

Comment:
s/ranges/originRanges/
s/split/failedRange/

Reply:
This change is unnecessary, `ranges` and `split` is straightforward enough.

Reply:
or `retryRange`

Reply:
I think it's good enough now. `retryRange` introduce ambiguity: the returned result may also called `retryRange`